loopy/tools.py

```diff
@@ -604,4 +612,128 @@ def intern_frozenset_of_ids(fs):
     return frozenset(intern(s) for s in fs)
 
 
-# vim: foldmethod=marker
+def _verify_generated_reproducer_is_equivalent(python_code, kernel, var_name):
+    reproducer_variables = {}
+    exec(python_code, reproducer_variables)
+    assert reproducer_variables[var_name] == kernel
+
+
+def kernel_to_python(kernel, var_name="kernel"):
+    """
+    Returns a :class:`str` of a python code that instantiates *kernel*.
+
+    :arg kernel: An instance of :class:`loopy.LoopKernel`
+    :arg var_name: A :class:`str` of the kernel variable name in the generated
+        python script.
+
+    .. note::
+
+        The implementation is partially complete and a :class:`AssertionError` is
+        raised if the if the returned python script does not exactly reproduce
+        *kernel*. Contributions are welcome to fill in the missing voids.
+    """
+    from mako.template import Template
+    from loopy.kernel.instruction import MultiAssignmentBase, BarrierInstruction
+
+    options = {}  # options: mapping from insn_id to str of options
+
+    for insn in kernel.instructions:
+        option = f"id={insn.id}, "
+        if insn.depends_on:
+            option += ("dep="+":".join(insn.depends_on)+", ")
+        if insn.tags:
+            option += ("tags="+":".join(insn.tags)+", ")
+
+        if isinstance(insn, MultiAssignmentBase):
+            if insn.atomicity:
+                option += "atomic, "
+        elif isinstance(insn, BarrierInstruction):
+            option += (f"mem_kind={insn.mem_kind}, ")
+        else:
+            pass
+
+        options[insn.id] = option[:-2]  # get rid of the trailing ", "
+
+    python_code = r"""
+    <%! import loopy as lp %>
+    import loopy as lp
+    import numpy as np
+    from pymbolic.primitives import *
```

Comment:
Maybe generate the imports as a separate string (or list of strings)?

Reply:
The routine `t_unit_to_python` now takes in an argument `return_preamble_and_body_separately` which does this. Thanks!